haystack_experimental/components/tools/tool_invoker.py

```diff
@@ -0,0 +1,547 @@
+        streaming_callback: Optional[StreamingCallbackT] = None,
+        *,
+        enable_streaming_callback_passthrough: bool = False,
+        max_workers: int = 4,
+        confirmation_strategies: Optional[dict[str, ConfirmationStrategy]] = None,
+    ):
+        """
+        Initialize the ToolInvoker component.
+
+        :param tools:
+            A list of tools that can be invoked or a Toolset instance that can resolve tools.
+        :param raise_on_failure:
+            If True, the component will raise an exception in case of errors
+            (tool not found, tool invocation errors, tool result conversion errors).
+            If False, the component will return a ChatMessage object with `error=True`
+            and a description of the error in `result`.
+        :param convert_result_to_json_string:
+            If True, the tool invocation result will be converted to a string using `json.dumps`.
+            If False, the tool invocation result will be converted to a string using `str`.
+        :param streaming_callback:
+            A callback function that will be called to emit tool results.
+            Note that the result is only emitted once it becomes available — it is not
+            streamed incrementally in real time.
+        :param enable_streaming_callback_passthrough:
+            If True, the `streaming_callback` will be passed to the tool invocation if the tool supports it.
+            This allows tools to stream their results back to the client.
+            Note that this requires the tool to have a `streaming_callback` parameter in its `invoke` method signature.
+            If False, the `streaming_callback` will not be passed to the tool invocation.
+        :param max_workers:
+            The maximum number of workers to use in the thread pool executor.
+            This also decides the maximum number of concurrent tool invocations.
+        :raises ValueError:
+            If no tools are provided or if duplicate tool names are found.
+        """
+        super(ToolInvoker, self).__init__(
+            tools=tools,
+            raise_on_failure=raise_on_failure,
+            convert_result_to_json_string=convert_result_to_json_string,
+            streaming_callback=streaming_callback,
+            enable_streaming_callback_passthrough=enable_streaming_callback_passthrough,
+            max_workers=max_workers,
+        )
+        self.confirmation_strategies = confirmation_strategies or []
+
+    def _handle_confirmation_strategies(
+        self, tool_calls: list[ToolCall], tool_call_params: list[dict[str, Any]]
+    ) -> tuple[list[ToolCall], list[dict[str, Any]], list[ChatMessage]]:
+        confirmed_tool_calls = []
+        confirmed_tool_call_params = []
+        tool_messages = []
+        for tool_call, params in zip(tool_calls, tool_call_params):
+            tool_name = params["tool_to_invoke"].name
+            if tool_name in self.confirmation_strategies:
```

Comment:
What happens if the tool has not a confirmation strategy?

Reply:
My intent was for tools without a strategy to just run as normal, but I see that's not reflected in the code. I can update that.